Gems/Atom/Feature/Common/Code/Source/Shadows/ProjectedShadowFeatureProcessor.cpp

```diff
@@ -299,50 +315,222 @@ namespace AZ::Render
         ShadowProperty& shadowProperty = m_shadowProperties.GetData(shadowPropertyIndex);
         shadowProperty.m_shadowId = shadowId;
 
-        AZ::Name viewName(AZStd::string::format("ProjectedShadowView (shadowId:%d)", shadowId.GetIndex()));
+        Name viewName(AZStd::string::format("ProjectedShadowView (shadowId:%d)", shadowId.GetIndex()));
         shadowProperty.m_shadowmapView = RPI::View::CreateView(viewName, RPI::View::UsageShadow);
 
         UpdateShadowView(shadowProperty);
+
+        if (m_primaryProjectedShadowmapsPass)
+        {
+            RPI::Ptr<ShadowmapPass> shadowmapPass = CreateShadowmapPass(shadowId.GetIndex());
+            m_shadowData.GetElement<ShadowPassIndex>(shadowId.GetIndex()) = shadowmapPass;
+            m_primaryProjectedShadowmapsPass->QueueAddChild(shadowmapPass);
+        }
     }
         
     void ProjectedShadowFeatureProcessor::OnRenderPipelineChanged([[maybe_unused]] RPI::RenderPipeline* renderPipeline,
             [[maybe_unused]] RPI::SceneNotification::RenderPipelineChangeType changeType)
     {
-        CachePasses();
+        if (changeType == RPI::SceneNotification::RenderPipelineChangeType::Removed)
+        {
+            // Check for cases where the pipeline containing the primary render passes has been removed, which means the pointers
+            // to those passes are no longer valid.
+            CheckRemovePrimaryPasses(renderPipeline);
+        }
+        if (changeType == RPI::SceneNotification::RenderPipelineChangeType::Removed || changeType == RPI::SceneNotification::RenderPipelineChangeType::PassChanged)
+        {
+            RemoveCachedPasses(renderPipeline);
+        }
+        if (changeType == RPI::SceneNotification::RenderPipelineChangeType::Added || changeType == RPI::SceneNotification::RenderPipelineChangeType::PassChanged)
+        {
+            CachePasses(renderPipeline);
+        }
+
+        // Check to see if the primary passes have changed, and if so removes the children from the old primary pass and creates them
+        // on the new primary pass. This is necessary if an earlier render pipeline adds or removes references to the shadow map
+        // passes, forcing this feature processor to change which pipeline it uses to render shadows for all pipelines in the scene.
+        UpdatePrimaryPasses();
```

Comment:
It seems UpdatePrimaryPasses() would set new primary passes. But the CachePasses() function is using primary passes to disable non-primary passes. With current order, CachePasses() is called before the new primary passes are set. Would it be a problem?

Reply:
Should be ok. The check is just comparing pointers in the esm case, the primary pass is never actually accessed. when there's some initialization going on and the primary esm pass hasn't yet been set, it'll disable it in `CachePasses()` then re-enable it later in `UpdateFilterParameters()`